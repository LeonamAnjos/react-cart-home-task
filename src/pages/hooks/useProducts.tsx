import React, { useEffect, useState } from "react";
import { Product } from "../structures/product.structures";

type UseProductsProps = {
  url: string;
};

const useProducts = ({ url }: UseProductsProps): Product[] => {
  const [products, setProducts] = useState<Product[]>([]);

  useEffect(() => {
    fetch(url)
      .then((response) => response.json())
      .then((response) => response.data)
      .then((data) => setProducts(data.products))
      .catch((err) => console.error(url, err)); // TODO: log it to remote
  }, [url]);

  return products;
};

export default useProducts;
