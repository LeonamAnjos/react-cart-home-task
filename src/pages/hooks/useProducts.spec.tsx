import { renderHook } from "@testing-library/react-hooks";
import { rest } from "msw";
import { setupServer } from "msw/node";
import useProducts from "./useProducts";

const url = "../../config/products.json";

const server = setupServer(
  rest.get(url, (req, res, ctx) => {
    return res(
      ctx.json({
        data: {
          products: [
            {
              id: 1,
              name: "Bike",
              minCoverage: 0,
              maxCoverage: 3000,
              risk: 0.3,
            },
            {
              id: 2,
              name: "Jewelry",
              minCoverage: 500,
              maxCoverage: 10000,
              risk: 0.05,
            },
          ],
        },
        error: {},
      })
    );
  })
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe("useProducts", () => {
  it("SHOULD fetch", async () => {
    const { result, waitForNextUpdate } = renderHook(() =>
      useProducts({ url })
    );

    await waitForNextUpdate();

    expect(result).toMatchSnapshot();
  });
});
