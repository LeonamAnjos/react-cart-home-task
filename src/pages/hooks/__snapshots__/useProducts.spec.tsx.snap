// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useProducts SHOULD fetch 1`] = `
Object {
  "all": Array [
    Array [],
    Array [
      Object {
        "id": 1,
        "maxCoverage": 3000,
        "minCoverage": 0,
        "name": "Bike",
        "risk": 0.3,
      },
      Object {
        "id": 2,
        "maxCoverage": 10000,
        "minCoverage": 500,
        "name": "Jewelry",
        "risk": 0.05,
      },
    ],
  ],
  "current": Array [
    Object {
      "id": 1,
      "maxCoverage": 3000,
      "minCoverage": 0,
      "name": "Bike",
      "risk": 0.3,
    },
    Object {
      "id": 2,
      "maxCoverage": 10000,
      "minCoverage": 500,
      "name": "Jewelry",
      "risk": 0.05,
    },
  ],
  "error": undefined,
}
`;
